simpeg_drivers/components/factories/misfit_factory.py

```diff
@@ -186,14 +197,58 @@ def create_nested_simulation(
         :param padding_cells: Number of padding cells around the local survey.
         """
         survey, indices, ordering = inversion_data.create_survey(
-            mesh=mesh, local_index=indices, channel=channel
+            mesh=global_mesh, local_index=indices, channel=channel
         )
         local_sim, mapping = inversion_data.simulation(
-            mesh,
+            global_mesh,
+            local_mesh,
             active_cells,
             survey,
             tile_id=tile_id,
             padding_cells=padding_cells,
         )
+        inv_type = inversion_data.params.inversion_type
+        if inv_type in ["fem", "tdem"]:
+            compute_em_projections(inversion_data, local_sim)
+        elif ("current" in inv_type or "polarization" in inv_type) and (
+            "2d" not in inv_type or "pseudo" in inv_type
+        ):
+            compute_dc_projections(inversion_data, local_sim, indices)
+        return local_sim, np.hstack(indices), ordering, mapping
+
+
+def compute_em_projections(inversion_data, simulation):
+    """
+    Pre-compute projections for the receivers for efficiency.
+    """
+    rx_locs = inversion_data.entity.vertices
+    projections = {}
+    for comp in "xyz":
+        projections[comp] = simulation.mesh.get_interpolation_matrix(
+            rx_locs, "faces_" + comp[0]
+        )
 
-        return local_sim, indices, ordering, mapping
+    for source in simulation.survey.source_list:
+        for receiver in source.receiver_list:
+            proj = 0.0
+            for ori, comp in zip(receiver.orientation, "xyz", strict=False):
+                if ori == 0:
+                    continue
+                proj += ori * projections[comp][receiver.local_index, :]
+            receiver.spatialP = proj
+
+
+def compute_dc_projections(inversion_data, simulation, indices):
+    """
+    Pre-compute projections for the receivers for efficiency.
+    """
+    rx_locs = inversion_data.entity.vertices
+    mn_pairs = inversion_data.entity.cells
+    projection = simulation.mesh.get_interpolation_matrix(rx_locs, "nodes")
+
+    for source, ind in zip(simulation.survey.source_list, indices, strict=False):
+        proj_mn = projection[mn_pairs[ind, 0], :]
+
+        if not np.all(mn_pairs[ind, 0] == mn_pairs[ind, 1]):
+            proj_mn -= projection[mn_pairs[ind, 1], :]
+        source.receiver_list[0]._Ps[simulation.mesh.n_cells] = proj_mn  # pylint: disable=protected-access
```

Comment:
We should take a look at all the private `SimPEG` attributes our codes use and make them public (and give them better names!).

Reply:
Yea I created spatialP for the EM's. I can expand here.